File: src/server/src/models/Post.ts
import { DataTypes, Model } from 'sequelize';
import db from '../db';
import { v4 as uuidv4 } from 'uuid';

class Post extends Model {
  declare id: typeof uuidv4;
  declare title: string;
  declare source: string;
  declare origin: string;
  declare description: string;
  declare contentType:
    | 'text/markdown'
    | 'text/plain'
    | 'application/base64'
    | 'image/png;base64'
    | 'image/jpeg;base64';
  declare content: string;
  declare image: string;
  declare categories: string[];
  declare count: number;
  declare published: Date;
  declare visibility: 'PUBLIC' | 'FRIENDS';
  declare unlisted: boolean;
}

Post.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      allowNull: false,
      primaryKey: true,
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    source: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    origin: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    description: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    contentType: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: 'text/plain',
      validate: {
        customValidator: value => {
          const enums = [
            'text/markdown',
            'text/plain',
            'application/base64',
            'image/png;base64',
            'image/jpeg;base64',
          ];
          if (!enums.includes(value)) {
            throw new Error('Not a valid option');
          }
        },
      },
    },
    content: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    image: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    categories: {
      type: DataTypes.ARRAY(DataTypes.STRING),
      allowNull: false,
    },
    count: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    published: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW,
    },
    visibility: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: 'PUBLIC',
      validate: {
        customValidator: value => {
          const enums = ['PUBLIC', 'FRIENDS'];
          if (!enums.includes(value)) {
            throw new Error('Not a valid option');
          }
        },
      },
    },
    unlisted: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false,
    },
  },
  {
    sequelize: db,
    modelName: 'Post',
    underscored: true,
  }
);

export default Post;

File: service/models/post.py
from django.db import models
from django import forms
from service.models.author import Author
import uuid
from datetime import datetime, timezone
from django.conf import settings

def get_current_date():
    datetime.now(timezone.utc)

class Category(models.Model):
    data = models.CharField(max_length=32, primary_key=True)

class Post(models.Model):
    _id = models.URLField(primary_key=True) #post id
    title = models.CharField(max_length=32)
    source = models.URLField()
    origin = models.URLField()
    description = models.TextField()

    MARKDOWN = "text/markdown"
    PLAIN = "text/plain"
    BASE64 = "application/base64"
    IMAGE_PNG = "image/png"
    IMAGE_JPEG = "image/jpeg"

    CONTENT_TYPES = (
        (MARKDOWN, "text/markdown"),
        (PLAIN, "text/plain"),
        (BASE64, "application/base64"),
        (IMAGE_PNG, "image/png"),
        (IMAGE_JPEG, "image/jpeg")
    )

    contentType = models.CharField(max_length=20, choices=CONTENT_TYPES, default=MARKDOWN)
    content = models.TextField()
    author = models.ForeignKey(Author, on_delete=models.CASCADE)

    categories = models.ManyToManyField(Category)

    published = models.DateTimeField(default=get_current_date)

    PUBLIC = "PUBLIC"
    FRIENDS = "FRIENDS"

    VISIBILITY_CHOICES = (
        (PUBLIC, "PUBLIC"),
        (FRIENDS, "FRIENDS")
    )

    visibility = models.CharField(max_length=7, choices=VISIBILITY_CHOICES)
    unlisted = models.BooleanField(default=False)

    def save(self, *args, **kwargs):
        self.published = datetime.now(timezone.utc)
        super(Post, self).save(*args, **kwargs)

    def toJSON(self):
        return {
            "type": "post",
            "title": self.title,
            "id": self._id,
            "source": self.source,
            "origin": self.origin,
            "description": self.description,
            "contentType": self.contentType,
            "content": self.content,
            "author": self.author.toJSON(),
            "categories": list(self.categories.values_list(flat=True)),
            "published": str(self.published),
            "visibility": self.visibility,
            "unlisted": self.unlisted
        }

    def toObject(self, json_object):
        self._id = json_object["id"] #this is a url, we need to get the last item of the url
        self.title = json_object["title"]
        self.id = json_object["id"]
        self.source = json_object["source"]
        self.origin = json_object["origin"]
        self.description = json_object["description"]
        self.contentType = json_object["contentType"]
        self.content = json_object["content"]
        self.author = Author().toObject(json_object["author"])
        self.categories = json_object["categories"]
        self.comments = json_object["comments"]
        self.published = json_object["published"]
        self.visibility = json_object["visibility"]
        self.unlisted = bool(json_object["unlisted"])

    @staticmethod
    def create_post_id(author_id, post_id=None): #uses the last uuid value from author id, and generates a custom post_id
        if not post_id:
            post_id = uuid.uuid4()

        author_uuid = author_id.rsplit('/', 1)[-1]
        return f"{settings.DOMAIN}/authors/{author_uuid}/posts/{post_id}"

    def __str__(self):
        return f"{self.title}, {self.author}, {self.published}"

